src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs

```diff
@@ -0,0 +1,215 @@
+                {
+                    string asyncMethodName = methodSymbol.Name + MandatoryAsyncSuffix;
+                    IEnumerable<IMethodSymbol> methodSymbols = semanticModel.LookupSymbols(
+                        context.Operation.Syntax.GetLocation().SourceSpan.Start,
+                        methodSymbol.ContainingType,
+                        asyncMethodName,
+                        includeReducedExtensionMethods: true)
+                        .OfType<IMethodSymbol>();
+
+                    string containingMethodName = "";
+                    if (context.ContainingSymbol is IMethodSymbol parentMethod)
+                    {
+                        containingMethodName = parentMethod.Name;
+                    }
+
+                    SyntaxNode invokedMethodName = context.Operation.Syntax;
+                    Location invokedMethodLocation = invocationOperation.Syntax.GetLocation();
+
+                    foreach (IMethodSymbol method in methodSymbols)
+                    {
+                        if (!method.IsObsolete()
+                            && HasSupersetOfParameterTypes(method, methodSymbol)
+                            && method.Name != containingMethodName
+                            && method.HasAsyncCompatibleReturnType())
+                        {
+                            // An async alternative exists.
+                            ImmutableDictionary<string, string>? properties = ImmutableDictionary<string, string>.Empty
+                                .Add(AsyncMethodKeyName, asyncMethodName);
+
+                            Diagnostic diagnostic = Diagnostic.Create(
+                                Descriptor,
+                                invokedMethodLocation,
+                                properties,
+                                invokedMethodName.ToString(),
+                                asyncMethodName);
+                            context.ReportDiagnostic(diagnostic);
+
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Determines whether the given method has parameters to cover all the parameter types in another method.
+        /// </summary>
+        /// <param name="candidateMethod">The candidate method.</param>
+        /// <param name="baselineMethod">The baseline method.</param>
+        /// <returns>
+        ///   <c>true</c> if <paramref name="candidateMethod"/> has a superset of parameter types found in <paramref name="baselineMethod"/>; otherwise <c>false</c>.
+        /// </returns>
+        private static bool HasSupersetOfParameterTypes(IMethodSymbol candidateMethod, IMethodSymbol baselineMethod)
+        {
+            return candidateMethod.Parameters.All(candidateParameter => baselineMethod.Parameters.Any(baselineParameter => baselineParameter.Type?.Equals(candidateParameter.Type) ?? false));
+        }
```

Comment:
This is a fine way to start for now.  Subsequently, though, we might want to consider augmenting it to handle common patterns between sync and async variants.  For example, a sync method might take `byte[] array, int offset, int count`, but the newer async variant might take `Memory<byte>`.